design-documents/graph-ql/framework/async-mutations.md

```diff
@@ -0,0 +1,281 @@
+# Asynchronous mutations
+
+## Purpose
+
+As a platform, Magento is going to evolve to support higher concurrency of operations.
+The best we can do to achieve this is the introduction of asynchronous commands for state modifying operations.
+As a result, the storefront application will be capable to process flow without waiting for a response.
+Such defer will allow server side to aggregate multiple requests and process them in the most efficient way we can achieve with the current application.
+The decision to use asynchronous operations requires changes in mutation interfaces design.
+
+## Schema for asynchronous mutation
+
+Let's consider an example of an asynchronous workflow by using with shopping cart mutations.
+
+### Setting a command 
+Actually, because we do not want to process command immediately mutations can not return results.
+Instead of the result, such mutation will return a code which will identify a command in execution queue.
+
+```graphql schema
+type Mutation {
+    createEmptyCart(cartMask: String!): OperationStateInterface!
+}
+interface OperationStateInterface {
+    identifier: String!
+}
+type OperationAccepted implements OperationStateInterface {
+     identifier: String!
+}
+```
+
+#### Request
+```graphql
+mutation {
+    reserveCart($cartMask: String!) {
+        identifier
+        __typename__
+    }
+}
+```
+
+#### Response
+```json
+{
+   "data":{
+      "operation":{
```

Comment:
Looks like there is no `operation` field in the schema currently, it should be added there. Otherwise "CreateEmptyCart(): identifier" sounds like it returns Cart Identifier.